cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteProgramRunDispatcher.java

```diff
@@ -78,23 +76,20 @@ public class RemoteProgramRunDispatcher implements ProgramRunDispatcher {
     .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
     .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
     .create();
-  private final Store store;
   private final ProgramRunnerFactory programRunnerFactory;
   private final RemoteTaskExecutor remoteTaskExecutor;
   private final ProgramRunnerFactory remoteProgramRunnerFactory;
-  private final RemoteExecutionTwillRunnerService remoteTwillRunnerService;
+  private final TwillControllerCreatorFactory twillControllerCreatorFactory;
 
   @Inject
   public RemoteProgramRunDispatcher(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
-                                    RemoteClientFactory remoteClientFactory, Store store,
+                                    RemoteClientFactory remoteClientFactory,
                                     ProgramRunnerFactory programRunnerFactory,
                                     @RemoteExecution ProgramRunnerFactory remoteProgramRunnerFactory,
-                                    @RemoteExecution TwillRunnerService twillRunnerService) {
-    this.store = store;
+                                    TwillControllerCreatorFactory twillControllerCreatorFactory) {
```

Comment:
The extra layer of TwillControllerCreator is added because we'll need to perform different steps to create the respective TwillControllers. For remote, we need to make an additional call to the store to fetch run record detail. Moreover, different functions of the respective TwillRunnerService implementations are being used.